### PR TITLE
Retrofit `IOContext` with `ErrorReportConfiguration`

### PR DESCRIPTION
**Pull request by JooHyukKim**
part of #1066 

~~(blocked by #1067)~~

<del>
### Notes

1. As suggested by [original PR 1043](https://github.com/FasterXML/jackson-core/pull/1043#issuecomment-1650737034), trying to minimize changes by

- Introduce "another" new constructor in 2.16(This will keep tests unchanged)
- Use `defaults()` method and not use new constructor in old constructors

2. After this PR merges, we can do clean ups by

- retrofitting old constructors with new constructor 
- Remove the first new constructor in Jackson 2.16 (one without `ErrorReportConfiguration`</del>

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/core/io/IOContext.java [hunk 1]
```diff
@@ -199,6 +233,16 @@ public StreamWriteConstraints streamWriteConstraints() {
...
     }
 
+    /**
+     * Returns : {@link ErrorReportConfiguration}, container for configuration values used when 
+     * handling errorneous token inputs. 
+     *
+     * @since 2.16
+     */
+    public ErrorReportConfiguration errorReportConfiguration() {
+        return _errorReportConfiguration;
```
Comment (pjfanning) [inline src/main/java/com/fasterxml/jackson/core/io/IOContext.java:243]: can this be called `getErrorReportConfiguration`? Jackson typically uses set/get method naming.

File: src/main/java/com/fasterxml/jackson/core/io/IOContext.java [hunk 2]
```diff
@@ -129,6 +136,8 @@ public class IOContext
      * @param managedResource Whether input source is managed (owned) by Jackson library
      *
      * @since 2.16
+     * @deprecated Since 2.16, use {@link #IOContext(StreamReadConstraints, StreamWriteConstraints, BufferRecycler, 
+     * ContentReference, boolean, ErrorReportConfiguration)} instead.
      */
     public IOContext(StreamReadConstraints src, StreamWriteConstraints swc, BufferRecycler br,
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/io/IOContext.java:142]: Could you also add `@Deprecated` here (for some reason Eclipse warns if that's missing when there's Javadoc `@deprecated`).

File: src/main/java/com/fasterxml/jackson/core/io/IOContext.java [hunk 3]
```diff
@@ -141,6 +150,30 @@ public IOContext(StreamReadConstraints src, StreamWriteConstraints swc, BufferRe
         _contentReference = contentRef;
         _sourceRef = contentRef.getRawContent();
         _managedResource = managedResource;
+        _errorReportConfiguration = ErrorReportConfiguration.defaults();
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/io/IOContext.java:153]: Maybe instead call:

    this(src, swc, br, contentRef, managedResource, ErrorReportConfiguration.defaults();

File: src/main/java/com/fasterxml/jackson/core/io/IOContext.java [hunk 4]
```diff
@@ -199,6 +233,16 @@ public StreamWriteConstraints streamWriteConstraints() {
         return _streamWriteConstraints;
     }
 
+    /**
+     * Returns : {@link ErrorReportConfiguration}, container for configuration values used when 
+     * handling errorneous token inputs. 
+     *
+     * @since 2.16
+     */
+    public ErrorReportConfiguration getErrorReportConfiguration() {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/io/IOContext.java:242]: Let's follow the New Jackson Naming convention :)

and drop "get", so `errorReportConfiguration()` (like `streamWriteConstraints()`)

**Comment by cowtowncoder** [comment]
Looks good, added couple of suggestions. You need to merge/rebase from 2.16, I made some changes hopefully easy enough to merge.

File: src/main/java/com/fasterxml/jackson/core/io/IOContext.java (hunk 4 above)
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/core/io/IOContext.java:242]: Ok, I notice @pjfanning requested `getXxx()` accessor, so it was "right way" at first. We are moving to reduce use of "get" in cases where there is no existing convention (for particular class) -- `get` and `set` method are to be used for general purpose properties/attributes, but not for configuration.
For time being there is obviously some inconsistency: for 3.0 more of get/set prefixes are being/have been removed. But there's no urgency for 2.x.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Ok, and then one important next step is to wire `JsonFactory.Builder` to allow configuration... right now we just get `ErrorReportConfiguration` defaults.

**Pull request merged**